tasks/task_database_cue.c

```diff
@@ -124,196 +301,650 @@ static int64_t get_token(intfstream_t *fd, char *token, uint64_t max_len)
-   uint8_t* boot_file;
-   int skip, frame_size, cd_sector;
-   uint8_t buffer[2048 * 2];
-   int is_mode1 = 0;
-
-   buffer[0]    = '\0';
-
-   if (intfstream_seek(fp, 0, SEEK_END) == -1)
-      return 0;
+   unsigned pos;
+   char raw_game_id[50];
+   char pre_game_id[50];
+   char hyphen = '-';
 
-   if (!sub_channel_mixed)
+   for (pos = 0; pos < 100000; pos++)
    {
-      if (!(intfstream_tell(fp) & 0x7FF))
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, raw_game_id, 12) > 0)
       {
-         unsigned int mode_test = 0;
+         raw_game_id[12] = '\0';
+         strncpy(pre_game_id, raw_game_id, 4);
+         pre_game_id[4] = '\0';
+         if (raw_game_id[4] == '_')
+         {
+            if (
+                  (string_is_equal(pre_game_id, "SCUS"))
+                  || (string_is_equal(pre_game_id, "SLUS"))
+                  || (string_is_equal(pre_game_id, "SLES"))
+                  || (string_is_equal(pre_game_id, "SCED"))
+                  || (string_is_equal(pre_game_id, "SLPS"))
+                  || (string_is_equal(pre_game_id, "SLPM"))
+                  || (string_is_equal(pre_game_id, "SCPS"))
+                  || (string_is_equal(pre_game_id, "SLED"))
+                  || (string_is_equal(pre_game_id, "SIPS"))
+                  || (string_is_equal(pre_game_id, "ESPM"))
+               )
+            {
+               raw_game_id[0] = raw_game_id[0];
+               raw_game_id[1] = raw_game_id[1];
+               raw_game_id[2] = raw_game_id[2];
+               raw_game_id[3] = raw_game_id[3];
+               raw_game_id[4] = hyphen;
+               raw_game_id[5] = raw_game_id[5];
+               raw_game_id[6] = raw_game_id[6];
+               raw_game_id[7] = raw_game_id[7];
+               raw_game_id[8] = raw_game_id[9];
+               raw_game_id[9] = raw_game_id[10];
+               raw_game_id[10] = '\0';
+
+               remove_spaces (game_id, raw_game_id);
+               append_multi_disc_suffix (game_id, filename);
+               return true;
+            }
+         }
+         else if (string_is_equal(pre_game_id, "LSP-"))
+         {
+            remove_spaces (game_id, raw_game_id);
+			game_id[10] = '\0';
+            append_multi_disc_suffix (game_id, filename);
+            return true;			
+         }
+      }
+      else
+         return false;
+   }
+   strcpy(game_id, "0000000000");
+   game_id[10] = '\0';
+   append_multi_disc_suffix (game_id, filename);
+   return true;
+}
 
-         if (intfstream_seek(fp, 0, SEEK_SET) == -1)
-            return 0;
+int detect_psp_game(intfstream_t *fd, char *game_id, const char *filename)
+{
+   unsigned pos;
 
-         intfstream_read(fp, &mode_test, 4);
-         if (mode_test != MODETEST_VAL)
-            is_mode1 = 1;
+   for (pos = 0; pos < 100000; pos++)
+   {
+      intfstream_seek(fd, pos, SEEK_SET);
+
+      if (intfstream_read(fd, game_id, 5) > 0)
+      {
+         game_id[5] = '\0';
+         if (
```

Comment:
Going to keep this a 4 byte string comparison as it will exactly match, where U or N would could be false positives.

Reply:
I wasn't suggesting removing the string comparisons. Since all of the string comparisons start with a U or N, you can avoid doing them if the next character is not a U or N.

Reply:
I see, first comparison checks U or N... if true then do the 4-5 byte comparison.  I'll put that in.

Reply:
Done